### PR TITLE
Generalize Fold[A] to Fold[A,B] (accumulator type);

### LINKED ISSUE #173: Generalize Array.fold operation
Currently it has following signature:
`def fold[A](Array[A], zero: A, (A,A) => A): A`
It can be generalized to the following:
`def fold[A,B](Array[A], zero: B, (B,A) => B): B`

### PR DESCRIPTION
**Pull request by greenhat**
Close #173

### CODE REVIEW

File: src/main/scala/sigmastate/lang/SigmaSpecializer.scala [hunk 1]
```diff
@@ -130,11 +130,11 @@ class SigmaSpecializer(val builder: SigmaBuilder) {
...
-    case Apply(Select(col,"fold", _), Seq(zero, Lambda(Seq((zeroArg, tZero), (opArg, tOp)), _, Some(body)))) =>
-      val taggedZero = mkTagged(zeroArg, tZero, 21)
-      val taggedOp = mkTagged(opArg, tOp, 22)
-      val body1 = eval(env ++ Seq(zeroArg -> taggedZero, opArg -> taggedOp), body)
-      Some(mkFold(col.asValue[SCollection[SType]], taggedZero.varId, zero, taggedOp.varId, body1))
+    case Apply(Select(col,"fold", _), Seq(zero, Lambda(Seq((accArg, tAccArg), (opArg, tOpArg)), _, Some(body)))) =>
+      val taggedAcc = mkTagged(accArg, tAccArg, 21)
+      val taggedOp = mkTagged(opArg, tOpArg, 22)
+      val body1 = eval(env ++ Seq(accArg -> taggedAcc, opArg -> taggedOp), body)
+      Some(mkFold(col.asValue[SCollection[SType]], taggedOp.varId, zero, taggedAcc.varId, body1))
```
Comment (greenhat) [inline src/main/scala/sigmastate/lang/SigmaSpecializer.scala]: @ergomorphic Could you please check if I get the intended parameter sequence right?

File: src/main/scala/sigmastate/utxo/transformers.scala [hunk 2]
```diff
@@ -180,46 +180,34 @@ case class ForAll[IV <: SType](input: Value[SCollection[IV]],
 }
 
 
-case class Fold[IV <: SType](input: Value[SCollection[IV]],
+case class Fold[IV <: SType, OV <: SType](input: Value[SCollection[IV]],
                              id: Byte,
-                             zero: Value[IV],
+                             zero: Value[OV],
                              accId: Byte,
```
Comment (greenhat) [inline src/main/scala/sigmastate/utxo/transformers.scala]: @ergomorphic I found `id` and  `accId` names and/or order confusing. Please confirm their correctness.

Comment (ergomorphic) [reply src/main/scala/sigmastate/utxo/transformers.scala]: @greenhat yes, but let's leave is they are. I'm working towards adding Function node in the nearest future, to have first class lambdas, so all this collection ops will be refactored.

File: src/main/scala/sigmastate/utxo/transformers.scala [hunk 3]
```diff
@@ -180,46 +180,34 @@ case class ForAll[IV <: SType](input: Value[SCollection[IV]],
...
-    Fold(input, 21, Constant(tT.upcast(0.toByte), tT), 22, Plus(TaggedVariable(22, tT), TaggedVariable(21, tT)))
-
-  def concat[T <: SType](input: Value[SCollection[SCollection[T]]])(implicit tT: T) = {
-    val tCol = SCollection(tT)
-    Fold[SCollection[T]](
-      input, 21, ConcreteCollection()(tT), 22,
-      Append(TaggedVariable(22, tCol), TaggedVariable(21, tCol)))
-  }
-}
-
```
Comment (ergomorphic) [inline src/main/scala/sigmastate/utxo/transformers.scala]: @greenhat please leave these methods.

File: src/test/scala/sigmastate/utxo/ErgoLikeInterpreterSpecification.scala [hunk 4]
```diff
@@ -151,14 +150,22 @@ class ErgoLikeInterpreterSpecification extends SigmaTestingCommons {
...
-          |  let outSumBytes = outBytes.fold(Array[Byte](), fun (arr1: Array[Byte], arr2: Array[Byte]) = arr2 ++ arr1)
+          |  let outSumBytes = outBytes.fold(Array[Byte](), fun (arr1: Array[Byte], arr2: Array[Byte]) = arr1 ++ arr2)
           |  let timePassed = HEIGHT > timeout
           |  notTimePassed && blake2b256(outSumBytes) == properHash || timePassed && sender
            }""".stripMargin).asBoolValue
 
       val prop = OR(
         AND(LE(Height, LongConstant(timeout)),
-          EQ(CalcBlake2b256(Fold.concat(MapCollection(Outputs, 21, ExtractBytesWithNoRef(TaggedBox(21))))),
+          EQ(CalcBlake2b256(
```
Comment (ergomorphic) [inline src/test/scala/sigmastate/utxo/ErgoLikeInterpreterSpecification.scala]: @greenhat please rollback this change

File: src/test/scala/sigmastate/utxo/ErgoLikeInterpreterSpecification.scala [hunk 5]
```diff
@@ -201,7 +208,13 @@ class ErgoLikeInterpreterSpecification extends SigmaTestingCommons {
     val propExp = AND(
       pubkey,
       GT(
-        Fold.sum[SLong.type](MapCollection(Outputs, 21, ExtractAmount(TaggedBox(21)))),
+        Fold(
```
Comment (ergomorphic) [inline src/test/scala/sigmastate/utxo/ErgoLikeInterpreterSpecification.scala]: @greenhat please rollback this change

**Comment by greenhat** [comment]
Fixed.

File: src/test/scala/sigmastate/serialization/generators/TransformerGenerators.scala [hunk 6]
```diff
@@ -52,9 +53,10 @@ trait TransformerGenerators {
     condition <- Gen.oneOf(TrueLeaf, FalseLeaf)
   } yield mkForAll(input, idByte, condition).asInstanceOf[ForAll[SInt.type]]
 
-  val foldGen: Gen[Fold[SInt.type]] = for {
+  val foldGen: Gen[Fold[SInt.type, SBoolean.type]] = for {
     input <- arbCCOfIntConstant.arbitrary
-  } yield Fold.sum(input)
```
Comment (ergomorphic) [inline src/test/scala/sigmastate/serialization/generators/TransformerGenerators.scala]: @greenhat  please leave `sum` method.

**Pull request merged**